core/src/banking_stage.rs

```diff
@@ -510,14 +524,36 @@ impl BankingStage {
         let num_to_commit = num_to_commit.unwrap();
 
         if num_to_commit != 0 {
-            bank.commit_transactions(
-                txs,
-                None,
-                &mut loaded_accounts,
-                &results,
-                tx_count,
-                signature_count,
-            );
+            let transaction_statuses = bank
+                .commit_transactions(
+                    txs,
+                    None,
+                    &mut loaded_accounts,
+                    &results,
+                    tx_count,
+                    signature_count,
+                )
+                .1;
+            if let Some(blocktree) = maybe_blocktree {
+                for (result, transaction) in transaction_statuses.iter().zip(batch.transactions()) {
+                    if Bank::can_commit(&result) && !transaction.signatures.is_empty() {
+                        let lamports_per_signature = bank
+                            .last_blockhash_with_fee_calculator()
```

Comment:
This doesn't look right.   The fee calculator should be from the transaction's recent_blockhash, not the latest blockhash in the bank